crates/bevy_core_pipeline/src/bloom/bloom.wgsl

```diff
@@ -1,138 +1,148 @@
-    o = o + (f + g + l + k) * div.y;
-    o = o + (g + h + m + l) * div.y;
-
-    return o;
+// [COD] slide 153
+fn sample_input_13_tap(uv: vec2<f32>) -> vec3<f32> {
+    let texel_size = 1.0 / vec2<f32>(textureDimensions(input_texture));
+    let x = texel_size.x;
+    let y = texel_size.y;
+
+    let a = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y + 2.0 * y)).rgb;
+    let b = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y + 2.0 * y)).rgb;
+    let c = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y + 2.0 * y)).rgb;
+
+    let d = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y)).rgb;
+    let e = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y)).rgb;
+    let f = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y)).rgb;
+
+    let g = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y - 2.0 * y)).rgb;
+    let h = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y - 2.0 * y)).rgb;
+    let i = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y - 2.0 * y)).rgb;
+
+    let j = textureSample(input_texture, s, vec2<f32>(uv.x - x, uv.y + y)).rgb;
+    let k = textureSample(input_texture, s, vec2<f32>(uv.x + x, uv.y + y)).rgb;
+    let l = textureSample(input_texture, s, vec2<f32>(uv.x - x, uv.y - y)).rgb;
+    let m = textureSample(input_texture, s, vec2<f32>(uv.x + x, uv.y - y)).rgb;
+
+#ifdef FIRST_DOWNSAMPLE
+    // [COD] slide 168
+    //
+    // The first downsample pass reads from the rendered frame which may exhibit
+    // 'fireflies' (individual very bright pixels) that should not cause the bloom effect.
+    //
+    // The first downsample uses a firefly-reduction method proposed by Brian Karis
+    // which takes a weighted-average of the samples to limit their luma range to [0, 1].
+    // This implementation matches the LearnOpenGL article [PBB].
+    var group0 = (a + b + d + e) * (0.125f / 4.0f);
+    var group1 = (b + c + e + f) * (0.125f / 4.0f);
+    var group2 = (d + e + g + h) * (0.125f / 4.0f);
+    var group3 = (e + f + h + i) * (0.125f / 4.0f);
+    var group4 = (j + k + l + m) * (0.5f / 4.0f);
+    group0 *= karis_average(group0);
+    group1 *= karis_average(group1);
+    group2 *= karis_average(group2);
+    group3 *= karis_average(group3);
+    group4 *= karis_average(group4);
+    return group0 + group1 + group2 + group3 + group4;
+#else
+    var sample = (a + c + g + i) * 0.03125;
+    sample += (b + d + f + h) * 0.0625;
+    sample += (e + j + k + l + m) * 0.125;
+    return sample;
+#endif
 }
 
-// Samples original using a 3x3 tent filter.
-//
-// NOTE: Use a 2x2 filter for better perf, but 3x3 looks better.
-fn sample_original_3x3_tent(uv: vec2<f32>, scale: vec2<f32>) -> vec4<f32> {
-    let d = vec4<f32>(1.0, 1.0, -1.0, 0.0);
+// [COD] slide 162
+fn sample_input_3x3_tent(uv: vec2<f32>) -> vec3<f32> {
+    // Radius. Empirically chosen by and tweaked from the LearnOpenGL article.
+    let x = 0.004;
+    let y = 0.004;
```

Comment:
Maybe this is the source of the “triplication” in @2x / retina resolutions? Would using the texel size here like above help?

Reply:
Just tried changing to `0.002` and it looks much nicer on retina, could we send the pixel ratio down to the shader, or base it on the texel size?

<img width="213" alt="image" src="https://user-images.githubusercontent.com/418473/219894915-0cfbf9e8-b62b-4e82-a0c1-efe1946f4a26.png">

Reply:
This seems to make it work consistently on both 1x and 2x scaling on macOS, we can tweak the `3.0` value if needed but it seems to roughly match the current `0.004` visually

```suggestion
    let texel_size = 1.0 / vec2<f32>(textureDimensions(input_texture));
    let x = texel_size.x * 3.0;
    let y = texel_size.y * 3.0;
```

Reply:
Amazing job! I'd been so annoyed by the weird rings and bloom stretching but couldn't point out what was causing it. I thought the rings were an artifact of the blending function which lead me on a whole journey at the end of which I built a [parametric lens scattering profile visualizer](https://starlederer.github.io/bloom/) (still useful), but I didn't know what to do with aspect ratio issues. Thank you very much!!

Reply:
Additional context: I chatted with @Elabajaba, @superdump, Jasmine (not sure on their GitHub handle) and @DGriffin91 yesterday [via Discord](https://discord.com/channels/691052431525675048/743663924229963868/1076600909133529188), looks like there's a trade-off in keeping the relative size of the bloom consistent across resolutions (for consistent look and artistic control) and avoiding these artifacts, we discussed multiple alternatives, including one where we only correct the aspect ratio but not the size (see code below) or using different sampling approaches with different tap configurations. Not sure if this is a “final” solution given that

```rust
    let texel_size = 1.0 / vec2<f32>(textureDimensions(input_texture));
    let aspect_ratio = texel_size.x / texel_size.y;
    let x = 0.004 * aspect_ratio;
    let y = 0.004;
```

Reply:
All of these still suffer from the weird "triplicated" artifacting btw.

Reply:
Perhaps a way to fix this is to downscale the view to 1920x1080 or some other fixed resolution, apply the bloom, and then upscale it back when compositing? Otherwise, we would need a large number of samples to make it look decent at higher resolution.

Reply:
agree with using the aspect ratio but i think it should be passed in a uniform as the aspect ratios are not consistent over all the downsampled levels. this is particularly noticable if we use `width.max(height)` to get the mip count instead of min

Reply:
looks like using a uniform fixes the triplication as well